src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1305,7 +1305,7 @@ private Solution AddDocument(Document document)
             CheckNotContainsDocument(document.Id);
             CheckContainsProject(document.Id.ProjectId);
 
-            return this.AddDocument(document.State);
+            return this.AddDocument((DocumentState)document.State);
```

Comment:
Then "Document.State" should be strongly typed to return DocumentState :)

Right now with your current approach you're making an assumption which could break if we change how Document works.  Have the type system actually enforce these invariants.